src/Components/Web.Extensions/src/InputFile/InputFile.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    /// <summary>
+    /// A component that wraps the HTML file input element and exposes a <see cref="Stream"/> for each file's contents.
+    /// </summary>
+    public class InputFile : ComponentBase, IInputFileJsCallbacks, IDisposable
+    {
+        private ElementReference _inputFileElement;
+
+        private IJSRuntime _jsRuntime = default!;
+
+        private IJSUnmarshalledRuntime? _jsUnmarshalledRuntime;
+
+        private InputFileJsCallbacksRelay? _jsCallbacksRelay;
+
+        [Inject]
+        private IServiceProvider ServiceProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the event callback that will be invoked when the collection of selected files changes.
+        /// </summary>
+        [Parameter]
+        public EventCallback<IFileListEntry[]> OnChange { get; set; }
+
+        /// <summary>
+        /// Gets or sets the maximum message size for file data sent over a circuit.
+        /// This only has an effect when using Blazor Server.
+        /// </summary>
+        [Parameter]
+        public int MaxMessageSize { get; set; } = 20 * 1024; // SignalR limit is 32K.
+
+        /// <summary>
+        /// Gets or sets the maximum internal buffer size for file data sent over a circuit.
+        /// This only has an effect when using Blazor Server.
+        /// </summary>
+        [Parameter]
+        public int MaxBufferSize { get; set; } = 1024 * 1024;
```

Comment:
As another attempt to clarify, how about something like `MaxUnreadMemoryBufferSize`? That is, if I'm correct that this represents the maximum amount of data that the .NET size will pull from JS and hold until the application code reads it.

Again, better suggestions would be welcome!

Reply:
Also to clarify some of these thoughts, for parameters/methods that we think people shouldn't use/override very often because we believe our defaults are usually right, we often like to given the deliberately long and obtuse names. Two reasons: (1) so that people who use them are less likely to make incorrect guesses about what they do, and (2) to make people feel less inclined to specify them all the time.

So for these two parameters I wouldn't be looking for the most concise name possible. On the contrary, since we think developers shouldn't usually set these properties, I'd prefer longer names!